src/sympc/protocol/beaver/beaver.py

```diff
@@ -4,52 +4,76 @@
 
+""" Those functions should be executed by the Trusted Party """
 
-def build_triples(
-    x: MPCTensor, y: MPCTensor, op_str: str
-) -> Tuple[MPCTensor, MPCTensor, MPCTensor]:
 
+def _get_triples(
+    op_str: str, nr_parties: int, a_shape: Tuple[int], b_shape: Tuple[int]
+) -> Tuple[Tuple[ShareTensor, ShareTensor, ShareTensor]]:
     """
     The Trusted Third Party (TTP) or Crypto Provider should provide this triples
     Currently, the one that orchestrates the communication provides those
     """
-    if op_str not in EXPECTED_OPS:
-        raise ValueError(f"{op_str} should be in {EXPECTED_OPS}")
+    a = ShareTensor(encoder_precision=0)
+    a.tensor = torch.zeros(a_shape, dtype=torch.long)
+    a_shares = MPCTensor.generate_shares(a, nr_parties, torch.long)
+
+    b = ShareTensor(encoder_precision=0)
+    b.tensor = torch.zeros(b_shape, dtype=torch.long)
+    b_shares = MPCTensor.generate_shares(b, nr_parties, torch.long)
 
-    shape_x = x.shape
-    shape_y = y.shape
+    cmd = getattr(operator, op_str)
 
-    session = x.session
+    c = ShareTensor(encoder_precision=0)
+    c.tensor = cmd(a.tensor, b.tensor)
+    c_shares = MPCTensor.generate_shares(c, nr_parties, torch.long)
 
-    a = ShareTensor(session=session)
-    a.tensor = torch.empty(size=shape_x, dtype=torch.long).random_(
-        generator=ttp_generator
-    )
+    return a_shares, b_shares, c_shares
 
-    b = ShareTensor(session=session)
-    b.tensor = torch.empty(size=shape_y, dtype=torch.long).random_(
-        generator=ttp_generator
-    )
 
-    cmd = getattr(operator, op_str)
+@register_primitive_generator("beaver_matmul")
+def get_triples_matmul(
+    *args: List[Any], **kwargs: Dict[Any, Any]
+) -> Tuple[List[ShareTensor], List[ShareTensor], List[ShareTensor]]:
+    return _get_triples("matmul", *args, **kwargs)
 
-    # Manually place the tensor and use the same session such that we do
-    # not encode the result
-    c = ShareTensor(session=session)
-    c.tensor = cmd(a.tensor, b.tensor)
 
-    a_sh = MPCTensor(secret=a, session=session)
-    b_sh = MPCTensor(secret=b, session=session)
-    c_sh = MPCTensor(secret=c, session=session)
+@register_primitive_generator("beaver_mul")
+def get_triples_mul(
+    *args: List[Any], **kwargs: Dict[Any, Any]
+) -> Tuple[List[ShareTensor], List[ShareTensor], List[ShareTensor]]:
+    return _get_triples("mul", *args, **kwargs)
+
+
+@register_primitive_generator("beaver_wraps")
+def count_wraps_rand(
+    nr_parties: int, shape: Tuple[int]
+) -> Tuple[List[ShareTensor], List[ShareTensor]]:
+    r = ShareTensor(encoder_precision=0)
+    r.tensor = torch.zeros(shape, dtype=torch.long)
+
+    r_shares = MPCTensor.generate_shares(r, nr_parties, torch.long)
+    wraps = count_wraps(r_shares)
+    for share in r_shares:
+        print(share)
+    print(wraps)
+
+    theta_r = ShareTensor(encoder_precision=0)
+    theta_r.tensor = wraps
+
+    theta_r_shares = MPCTensor.generate_shares(theta_r, nr_parties, torch.long)
```

Comment:
Hmm...maybe we can add a parameter, for the moment when we do the MPCTensor we also send the shares to the parties.
Using, ```MPCTensor.generate_shares``` would only split the secret, but the shares are kept local.

We can have:
```
x = ShareTensor(data = y, encoder_precision=0)
x_shares = MPCTensor.generate_shares(x, ...)
```

Or, I was thinking to have something like:
```x = MPCTensor.generate_shares(x)``` <--- and here to default the encoder_precision to 0

Do you have any ideas in mind? or are the ones above more appealing?

Reply:
Currently, I replaced this with:
```
x = ShareTensor(data = y, encoder_precision=0)
x_shares = MPCTensor.generate_shares(x, ...)
```

I will push my changes this evening :D

Reply:
Ok cool, it's better!
But there is still this thing where I think "Ok I have generated a share (because I used "ShareTensor"), and then I use this in generate_shares ?!" Maybe something more direct like you suggested would be more readable for the new readers like me :D 
```python
x_shares = MPCTensor.generate_shares(x, encoder_precision=0, ...)
```

Reply:
I opened an issue for [it](https://github.com/OpenMined/SyMPC/issues/46) :( - I think this PR is getting very very massive.